Ucode early and late loading plus build cdboot locally.

01micko opened the pull request:
See #2344 for all the latest discussion

To DO

- [x] integrate into `3builddistro`
- [x] Wait for updates to `frugalpup`
- [x] test early loading with other boot ~managers~ loaders like ~refind, grub,~ grub4dos. `grub4dos` works - with `errorcheck off` in `menu.lst` stanza and calling both ucode.cpio and initrd.gz on the `initrd` line. 
- [x] basic auditing - eg record when files are built; check if new ones are available.
- [x] ~get individual blob versions for Intel and record them so not downloaded twice. AMD is more problematic but they haven't been updated in a while.~ not worth the effort on my side or the end user side as it involves more grokking out from downloaded files which can be slow on on a poor connection
- [x] update code in the woof iso routine
- [x] update 'full' install routine

Commit:
add file /usr/sbin/latest_microcode.sh re #2344

a symlink is made in /support so that we can call it from woof.

This script builds the ucode.cpio file containing the microcode for intel and amd

Commit:
Update bootmanager to support microcode building and loading

- has a few conditional calls for relevant exclusions, like cpu arch, pupmode etc

rizalmart committed:
create file /etc/init.d/rc.ucode re #2344 , closes #2316 , closes #2317 due to conflicts

Commit:
add file get_ucode.sh for late loading of microcode re #2344 #2316 #2317

downloads microcode specific to cpu according to /proc/cpuinfo

Commit:
add micro2pup.sh re #2344

This is a gtkdialog GUI that is called from bootmanager
It handles whether you want to update or install eraly and late microcode loading

Commit:
add AMD and Intel icons for micro2pup.sh GUI

These are hand drawn in code and a different enough from originals to avoid a law suit!

01micko commented:
Here are some screenies of the respective GUI

Bootmanager

![bootmanager](https://user-images.githubusercontent.com/1019119/123499976-ee2b4000-d67d-11eb-9f71-c77ace4895a5.png)


My Ryzen box with ucode.cpio loaded

![micro2pup_amd](https://user-images.githubusercontent.com/1019119/123499988-04d19700-d67e-11eb-9475-3801e751f0ef.png)

My Intel box (i686)

![micro2pup_intel](https://user-images.githubusercontent.com/1019119/123499998-187cfd80-d67e-11eb-8b18-bc6ce5c51e05.png)


(woops - just let it slip that I have slacko32-8.0 in development! :laughing: )

EDIT:

I haven't forgotten other architectures too.

Raspi shots follow (gtk2)

Bootmanager

![raspi-bootmanager](https://user-images.githubusercontent.com/1019119/123500209-ba511a00-d67f-11eb-8e1e-e386e00550af.png)


micro2pup

![micro2pup](https://user-images.githubusercontent.com/1019119/123500218-c8069f80-d67f-11eb-8c1f-e20540aee7fc.png)

Commit:
fix typo in latest_microcode.sh

thanks to @jamesbond3142

gyrog commented:
1  In 'latest_microcode.sh' you store it in '/mnt/home/$PSUBDIR',
this might not be a great idea, it might be better to extract the actual save directory from $PUPSAVE.
or save it in the / filesystem, like' /usr/etc/' or somewhere.
Current Puppy supports having a save location that has no relation to the install location,
and $PSUBDIR defines the install location.

2 In 'init.d/rc.ucode' you attempt to download microcode;
On my machines, I don't have a working network until several seconds after the desktop appears,
so even thinking of doing network stuff in 'init.d' seems extremely optimistic.

3 FrugalPup already coded, 'frugalpup-bootentry' change, and tested Ok.
It will insert the "if" statement into every grub2 boot entry it generates, quite harmless if the microcode initrd file does not exist in the install directory.
Just the resolution of an unrelated issue before release of v34.

01micko commented:
> 1 In 'latest_microcode.sh' you store it in '/mnt/home/$PSUBDIR',
> this might not be a great idea, it might be better to extract the actual save directory from $PUPSAVE.
> or save it in the / filesystem, like' /usr/etc/' or somewhere.
> Current Puppy supports having a save location that has no relation to the install location,
> and $PSUBDIR defines the install location.
> 

That is fully intentional. `ucode.cpio` is essentially an 'initial ram disk' and is supposed to be right next to `initrd.[gx]z`. Maybe I'm missing your point. :thinking: .. But maybe you are right. Feel free to offer advice.

> 2 In 'init.d/rc.ucode' you attempt to download microcode;
> On my machines, I don't have a working network until several seconds after the desktop appears,
> so even thinking of doing network stuff in 'init.d' seems extremely optimistic.
> 

There is a `check_route()` function that waits 30 seconds (approx- `sleep 1` is not exactly a second). If it's not enough the script fails. It can be called later like any other service script, and it only tries to download if the files aren't there. (that's the intention anyway - and this is for late loading too - nothing to do with creating the `ucode.cpio`)

> 3 FrugalPup already coded, 'frugalpup-bootentry' change, and tested Ok.
> It will insert the "if" statement into every grub2 boot entry it generates, quite harmless if the microcode initrd file does not exist in the install directory.
> Just the resolution of an unrelated issue before release of v34.

Great! Sounds good to me. release when you are comfortable.. that is why this PR is marked 'Draft' :) - so reviewers can pick it to bits and we end up with better code.

gyrog commented:
The problem is that '/mnt/home/$PSUBDIR' is in fact meaningless, '/mnt/home' is a symbolic link to the mount point for the save partition, '$PSUBDIR' is the sub-directory for the install location.
We've gotten away with this sort of thing, for many years because in a default frugal install, the directory that contains the "puppy...sfs" and the directory that contains the savefolder are one and the same.
BUT, this is not necessarily so, the save location can be on a different drive, in a sub-directory that is not the same as the one used for the install directory.

To write the microcode beside the "initrd.gz", beside the "puppy...sfs", you need to parse $PUPSFS.

But then we have the second part of the problem, the install directory could be on an ntfs partition, that is the Windows C: drive. In this case the user needs to be very aware that they are writting a file to that drive, before it happens. 
There is no way we can check that "Fast Startup" is in fact de-activated in Windows, so we can't be sure.
(The save location could easily be some where else, on a Linux partition.)

gyrog commented:
I guess I'm suggesting that copying the microcode initrd file into the install directory of a particular Puppy should be a separate deliberate step.
I'm also thinking of folk who have multiple Puppies, (I currently have 20), I would want to download the latest microcode for my CPU once, then copy it to the individual install directories when and if I so choose.

01micko commented:
> I guess I'm suggesting that copying the microcode initrd file into the install directory of a particular Puppy should be a separate deliberate step.
> I'm also thinking of folk who have multiple Puppies, (I currently have 20), I would want to download the latest microcode for my CPU once, then copy it to the individual install directories when and if I so choose.

No problem , I can pop a message that says 'ucode.initrd is in /tmp. Do you want to install it? [yes button] [no button]' perhaps with suggestions that they may want to store it and be careful if the partition is ntfs, however using this messes with my preferred method of creating a config file with a time stamp. Maybe I'll append the time stamp the the cpio archive? Then tailing it can retrieve the time stamp, but I'll have to test whether appending it breaks the archive or not.

> you need to parse $PUPSFS.

Will do.

This should be ok;
```
~$ PUPINSTALL=/initrd${PUP_HOME}${PUPSFS##*,}
~$ echo ${PUPINSTALL%/*} 
/initrd/mnt/dev_save/pups/slacko64-8.2.1

```

gyrog commented:
@01micko,
Sorry, but it won't do, again '/initrd/$PUP_HOME' is the mountpoint of the save partition, not the install partition. You would have to extract the partition name ${PUPSFS%%,*} and then find it's mountpoint, if it is mounted, if it's not, you would need to mount it.
In pupmodes 5,77,and 66, nothing is mounted by default, in pupmode=12,13, the save partition is always mounted "rw", but the install partition is mounted by default only if it happens to be the same as the save partition. 

I would suggest that rather than trying to put it in the install directory of the current running Puppy, your "download/chek-for latest" code gets a "microcode-storage-directory" from the user. Then if a 'ucode.cpio' already exists, you can directly check it's "modify" date and time, or store a config file beside it. And you have somewhere permanent to store the file.
Copying this microcode initrd to a particular Puppy install directory, could be a separate process.

01micko commented:
Ok @gyrog installing is optional now

```sh
~$ latest_microcode.sh -h
latest_microcode.sh-0.2

 latest_microcode.sh -h|--help : show this message
 latest_microcode.sh amd : build AuthenticAMD.bin microcode
 latest_microcode.sh intel : build GenuineIntel.bin microcode
 latest_microcode.sh [amd|intel|b|''] noinstall : build microcode, DON'T install
 latest_microcode.sh ucode-r : get the existing microcode release date and exit
 latest_microcode.sh remote-r : get the remote microcode release date and exit
 latest_microcode.sh b | OR with no arguments build both AMD and Intel microcode
```

The ` latest_microcode.sh $arg noinstall` command achieves this where are is amd, intel, '', or b (both).

The GUI calls this and the result is an option window after the `ucode.cpio` is built 

![micro2pup_install](https://user-images.githubusercontent.com/1019119/123530795-c3a2bb00-d741-11eb-9f50-8d3589bca01c.png)


Also the version is recorded _inside_ `ucode.cpio` and if needed is extracted. The GUI does this too calling `latest_microcode.sh ucode-r` and the remote versions are called too with `latest_microcode.sh remote-r`. That way recommendations in the GUI are made according to what's installed and what the remote versions are.

Committing shortly.

01micko commented:
So, now if the ${PUPSFS%%,*} isn't mounted then a mount point is made and it is mounted and ucode.cpio copied if 'install' is selected on cli or if 'yes' selected in GUI.

Commit:
Update latest_microcode.sh and micro2pup.sh GUI re #2344

Many thanks to @gyrog we now don't install by default and we find te correct location (hopefully) of where inird.[gx]z resides.

gyrog commented:
Just 1 simple question:
Approximately, how big is the download for microcode?

01micko commented:
4MB

Most of it is Intel. The one one for Late loading only gets 1 firmware so it's small.

If you choose only AMD it is 30K

Commit:
Merge branch 'testing' into ucode

Commit:
Merge branch 'testing' into ucode

Commit:
integrate latest_microcode.sh into 3builddistro

@dimkr - didn't touch dpup, I'll let you decide

Commit:
complete the 3builddistro integration

dimkr commented:
Is it enough to copy ucode.cpio to /mnt/home? If /mnt/home is the partition where the Puppy SFSs are, this won't work with UEFI because this file needs to be in the ESP.

dimkr commented on woof-code/rootfs-skeleton/usr/sbin/micro2pup.sh:
```diff
@@ -0,0 +1,252 @@
+MSG3=$(gettext 'This program does not support a Virtual Machine.')
+MSG4=$(gettext 'The puppy home drive is read only, however if saving session late loading is supported.')
+MSG5=$(gettext 'Your CPU vendor is unsupported.')
+PUPDIR=/initrd${PUP_HOME}${PUPSFS##*,}
```
What about 'full' installations? The fixes I'll need to do for `frugalify`/PUPMODE=6 images built in https://github.com/dpupos/woof-CE are the same or almost the same as those needed to support early microcode loading in 'full' installations.

01micko replied:
Full installs put `vmlinuz` in `/boot` IIRC, so if `ucode.cpio` is there and bootloader calls it shouldn't be an issue.

jamesbond3142 commented:
> Is it enough to copy ucode.cpio to /mnt/home? If /mnt/home is the partition where the Puppy SFSs are, this won't work with UEFI because this file needs to be in the ESP.

@dimkr The only constraint is that the bootloader need to be able to find and load it. If the existing bootloader can only load from ESP, then, yes, of course it has to be in ESP, but otherwise it can be anywhere.

> What about 'full' installations

Same, the bootloader still have to load it, and then it will automatically get used. My tests indicated that if you have "root=" line, full-install functionality will be used (the kernel won't attempt to run /init from initramfs if "root=" is specified).
For frugalify, if I get it right, you just need to specify "init=/sbin/frugalify" (to override the default /sbin/init), right? Everything else will be as full-install setup.

dimkr commented:
> For frugalify, if I get it right, you just need to specify "init=/sbin/frugalify" (to override the default /sbin/init), right? Everything else will be as full-install setup.

You need to specify `root` and `init=/init`, because the frugalify executable sits alongside the SFSs. The kernel needs to mount the partition and run frugalify, which chroots into an overlayfs of SFSs at /.

After frugalify runs, Puppy behaves sort of like a full installation (it doesn't know it runs under a union file system) but with stuff like /initrd and /mnt/home (bind mounts created by frugalify to simulate a 'frugal' installation) present.

If I understand correctly, initrd.gz is loaded to memory by the boot loader, so ucode.cpio must be in the same partition as the boot loader, and it's neither /initrd nor /mnt/home, the two bind mounts to the `root` partition, created by frugalify.

Commit:
Merge branch 'testing' into ucode

gyrog commented:
With current Puppy uefi booting, neither the 'initrd.gz' nor the 'vmlinuz' are in  the ESP,
so I don't expect any problems with 'ucode.cpio' not being in the ESP.

dimkr commented:
> With current Puppy uefi booting, neither the 'initrd.gz' nor the 'vmlinuz' are in the ESP,
> so I don't expect any problems with 'ucode.cpio' not being in the ESP.

Thanks @gyrog, didn't know that. I use efilinux in my builds, so vmlinuz is alongside bootx64.efi. There's no initrd.gz with frugalify, but I'll need to put ucode.cpio there.

Commit:
add UCODE_EXEC=../support/latest_microcode.sh to dpup configs

re 6cfb509

Commit:
Merge branch 'testing' into ucode

dimkr committed:
remove microcode packages from dpup

Commit:
Merge pull request #2365 from puppylinux-woof-CE/dpup-ucode

Remove microcode packages from dpup

Commit:
Merge branch 'testing' into ucode

Commit:
Merge branch 'testing' into ucode

Commit:
ucode.cpio support for slacko-8.2.2

dimkr commented on woof-code/3builddistro:
```diff
@@ -561,9 +561,17 @@ if [ ! "$CROSFLAG" ]; then
 	case $WOOF_TARGETARCH in x86*)
-		cp -fv ../initrd-progs/0initrd/README.txt build/ ;;
+		cp -fv ../initrd-progs/0initrd/README.txt build/ 
+		# ucode.cpio
```
This part should be outside of the `if [ ! "$CROSFLAG" ]; then`

01micko replied:
Ok in #2369 I see

dimkr commented on woof-code/rootfs-skeleton/usr/sbin/bootmanager:
```diff
@@ -346,13 +391,38 @@ else
 	SFSFRAME_XML="$(generic_xml)" #--
 fi
+case $KARCH in
+	x86_64|i?86|amd64);;
```
It's always i?386 or x86_64, never amd64 AFAIK

01micko replied:
Yeah we don't use $DEBARCH

dimkr committed:
add support for early microcode loading in Chrome OS style images

Commit:
Merge pull request #2369 from puppylinux-woof-CE/cros-ucode

Add support for early microcode loading in Chrome OS style images

Commit:
Merge branch 'testing' into ucode

Commit:
Merge branch 'testing' into ucode

Commit:
allow early ucode loading in full install

Commit:
Udate cdboot for microcode loading

- also fixed some typos and added new grub4dos splash image
- https for ibiblio

Commit:
Merge branch 'testing' into ucode

Commit:
remove amd64 form case in bootmanager

Commit:
Make late loading of microcode opt-in not opt-out

dimkr commented on woof-code/rootfs-skeleton/etc/init.d/rc.ucode:
```diff
@@ -0,0 +1,98 @@
+# if not download it and load it if possible
+# GPLv2
+
+[ -e "$HOME/.ucode.rc" ] && . $HOME/.ucode.rc # written by micro2pup.sh
```
is HOME set when running the init script? Why not /etc or ~?

01micko replied:
Fixed.

Commit:
HOME isn't set this early; change to /etc

gyrog commented:
The scripts in '/etc/init.d' are NOT run by the 'init' script.
These are the "services", $HOME is set by 'rc.sysinit', well before the scripts in '/etc/init.d' are run.

Commit:
Merge branch 'testing' into ucode

Commit:
new package grub4dos_u

This is an interim package until it gets wider testing.
It is basically the grub4dos package patched to allow loading of ucode.cpio
next to initrd.gz

Commit:
let Packages-puppy-noarch-official download from branch

relieves hard coding to 'testing'

Commit:
Merge branch 'testing' into ucode

01micko commented:
@gyrog ~I think~ the `ucode.cpio` isn't getting copied to the `initrd.gz` location in `frugalpup`, if it exists.

gyrog commented:
The only changes I made to FrugalPup is for it to generate boot-entires that will load the ucode.cpio file as an initrd if it exists in the installation directory.
Since the presence or absence of the ucode.cpio file is the trigger, surely it should be the responsibility of the user to "move" it there.
I must have missed the memo indicating that FrugalPup would be expected to do anything more.

01micko commented:
My intention was to issue a separate PR for `cdboot` but it is tightly integrated in to the function of this PR so doing it here - re #2391

Commit:
Merge branch 'testing' into ucode

Commit:
Don't show 'microcode' tab in bottmanager on non-x86 arches

Makes more sense than blurbing on about it in a non-functional GUI!

Commit:
clean up redundant stuff in /usr/share/boot-dialog

- removed 8k and added 5k
- only thing added from `cdboot` is the font.pf2 file (5k)
- nicer splash.jpg (same size)

Commit:
make UEFI cdboot locally built using local resources

- use frugalpup grub2 signed binaries - not debians
- create menus on the fly
- support 64 bit UEFI implementations by default and optionally 32 bit UEFI
- closes #2391

Commit:
Adjust all necessary _00build.conf files

- this just adds a comment block letting the builder know that they can support 32 bit UEFI implementations as well as 64 if the option is set

01micko commented:
lol .. `bottmanager` - wish I had one!

Anyway, almost ready.

Commit:
remove redundant grub2_efi package from all PKGS_SPECS_TABLE vars

No need when using more up to date and signed frugalpup binaries

Commit:
Merge branch 'testing' into ucode

Commit:
Merge branch 'testing' into ucode

Commit:
restore LICK windows installer documents in iso image

01micko commented:
I think this PR is complete. If @dimkr @gyrog @jamesbond3142 can take a look to see if I missed anything that would be good. I intend to merge in about 24 hours.

Commit:
Merge branch 'testing' into ucode

Commit:
Merge branch 'testing' into ucode
